src/Dapr.Actors/Runtime/ActorStateManager.cs

```diff
@@ -3,26 +3,31 @@
 // Licensed under the MIT License.
 // ------------------------------------------------------------
 
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Dapr.Actors.Resources;
+
 namespace Dapr.Actors.Runtime
 {
-    using System;
-    using System.Collections.Generic;
-    using System.Globalization;
-    using System.Threading;
-    using System.Threading.Tasks;
-    using Dapr.Actors.Resources;
-
-    internal sealed class ActorStateManager : IActorStateManager
+    internal sealed class ActorStateManager : IActorStateManager, IActorContextualState
     {
         private readonly Actor actor;
         private readonly string actorTypeName;
-        private readonly Dictionary<string, StateMetadata> stateChangeTracker;
+        private readonly ConcurrentDictionary<string, Dictionary<string, StateMetadata>> contextualStateChangeTracker;
```

Comment:
This approach has a pretty serious flaw because you need someone to **remove** the entries from the dictionary. If you think about a very long running application you'd accumulate millions of entries here just sitting in memory.

I'd suggest something different. Instead of:

```C#
private static AsyncLocal<string> context = new AsyncLocal<string>();
```

```C#
private static AsyncLocal<(string, Dictionary<string, StateMetadata>)> context = new AsyncLocal<(string, Dictionary<string, StateMetadata>)>();
```

Reply:
That's a much better way of doing it :) I was thinking about if I needed to remove the dictionary during a clear/reset but this removes that concern.

Reply:
You'd be surprised what can be a security bug 😉